Add RwLockReadGuard::map for transforming guards to contain sub-borrows.

reem opened the pull request:
<!-- homu-ignore:start -->
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/pull/30834)*
<!-- homu-ignore:end -->

This is very useful when the RwLock is synchronizing access to a data
structure and you would like to return or store guards which contain
references to data inside the data structure instead of the data structure
itself.

reem committed:
Add guard map methods for transforming guards to contain sub-borrows.

This is very useful when the lock is synchronizing access to a data
structure and you would like to return or store guards which contain
references to data inside the data structure instead of the data structure
itself.

rust-highfive commented:
r? @alexcrichton

(rust_highfive has picked a reviewer for you, use r? to override)

reem commented:
I've opened this without an accompanying RFC since it's a simple, unstable API. Let me know if adding this as unstable would require an RFC and I can write one up.

reem commented:
For reference, my current use case is a type which contains a `RwLock<HashMap<K, V>>` and I have an API where I would like to return `RwLockReadGuard<V>`, which is currently impossible. Without this API, I would have to define a new type and look up the value on every dereference of that type.

huonw commented:
I would think these APIs should match the similar APIs on the `RefCell` guards, e.g. http://doc.rust-lang.org/std/cell/struct.Ref.html has `map` and `filter_map`.

NB. those have the functions as associated functions, not methods: we try to avoid methods on generic smart-pointers/`Deref` types because they interfere with method calls to the internal one. The method form would actually be a breaking change, since, for instance, if `x: RwLock<Option<...>>` then `let y = x.read(); let z: Option<()> = y.map(|_| ())` compiles, but this PR would make `z` try to be assigned an `RwLockReadGuard<()>`.

Also, `MutexGuard` and `RwLockWriteGuard` could probably have them.

(I'm in favour of this.)

reem commented:
Note: I have prepped a temporary fork of RwLock to get this API on stable for my personal use which I'll keep around until these changes become stabilized. (We have a rust in prod usecase that depends on this feature and we would like to compile on stable)

Agreed with all of @huonw's comments, will update.

sfackler commented:
If we're going to add this (which I have no real problem with), we should also stick this on `MutexGuard` and `RwLockWriteGuard` in this PR.

reem commented:
Can do.

reem commented:
Hmm, I am looking at this code again and I think there is a subtle bug - the original guard should be `mem::forget`ed so as to avoid prematurely releasing the read lock.

rphmeier commented:
I think this is a definite improvement to the RwLock API, having found myself wanting something like this in my own code and being forced to approximate with unsafe code. I also agree with @sfackler that this should be added to the guards of various synchronization primitives, although the situation gets slightly hairier with `RwLockWriteGuard` and `MutexGuard`, since you have transformations from both `&mut T -> &mut U` as well as `&mut T -> &U`.

In particular, I wonder whether it is possible or makes sense to actually downgrade an `RwLockWriteGuard` to an `RwLockReadGuard` in the `&mut T -> &U` case, since it seems like overkill to continue to hold exclusive access if no further mutation is going to occur for the duration of the borrow.

sfackler commented:
@rphmeier pthread rwlocks don't support a downgrade operation, unfortunately, so we wouldn't be able to support a `&mut T -> &U` method on unix-y platforms.

rphmeier commented:
@sfackler Ah, that's unfortunate. At the very least, `&mut T -> &U` transformations can still be supported, they'll just have to continue holding exclusive access via a `RwLockWriteGuard`.

arthurprs commented:
This is very cool, I wanted something similar while writing floki.

sfackler commented:
I believe we're good to go with this if the same method gets added to `MutexGuard`, right @alexcrichton?

alexcrichton commented:
Ah yes sorry I meant to get to this sooner. We discussed this in the libs meeting and decided to move forward. We'd like to add this to the following types, though:

* `MutexGuard`
* `RwLockReadGuard`
* `RwLockWriteGuard`

Also feel free to hook this up to this tracking issue: https://github.com/rust-lang/rust/issues/27746

reem commented:
Updated the PR to include the method as an associated function for `RwLockReadGuard`, `RwLockWriteGuard` and `MutexGuard`, also added specific tests of poisoning behavior for `RwLockWriteGuard` and `MutexGuard`, where they are relevant.

reem commented:
I currently added the methods as unstable, but could change the PR to land them as insta-stable if that is desirable.

sfackler commented:
I think we'll land them as unstable but stabilize at the same time as `RefCellGuard::map`.

reem commented:
Ok, I changed them to use a common feature `guard_map`.

alexcrichton commented on src/libstd/sync/rwlock.rs:
```diff
@@ -417,10 +417,37 @@ impl<'rwlock, T: ?Sized> RwLockReadGuard<'rwlock, T> {
+               reason = "recently added, needs RFC for stabilization",
+               issue = "0")]
+    pub fn map<U: ?Sized, F>(this: Self, cb: F) -> RwLockReadGuard<'rwlock, U>
+    where F: FnOnce(&'rwlock T) -> &'rwlock U {
```
Stylistically where clauses like this in the standard library tend to be formatted like:

```rust
pub fn map<F>(...) -> T
    where F: ...
{
    // ...
}
```

alexcrichton commented on src/libstd/sync/rwlock.rs:
````diff
@@ -417,10 +417,37 @@ impl<'rwlock, T: ?Sized> RwLockReadGuard<'rwlock, T> {
+    /// Applies the supplied closure to the data, returning a new lock
+    /// guard referencing the borrow returned by the closure.
+    ///
+    /// ```rust
````
Could you add a `# Examples` heading above this as well?

alexcrichton commented on src/libstd/sync/rwlock.rs:
```diff
@@ -434,13 +461,52 @@ impl<'rwlock, T: ?Sized> RwLockWriteGuard<'rwlock, T> {
+    where F: FnOnce(&'rwlock mut T) -> &'rwlock mut U {
+        let new_data = unsafe {
+            let data: &'rwlock mut T = &mut *this.__data.get();
+            mem::transmute::<&'rwlock mut U, &'rwlock UnsafeCell<U>>(cb(data))
```
The strategy above of changing the stored type to just `T` seemed to work out well, could that be done here to avoid the transmute?

reem replied:
I'll change both `RwLockWriteGuard` and `MutexGuard` to store `&mut T`s.

reem committed:
Change MutexGuard and RwLockWriteGuard to store &mut T not &UnsafeCell<T>

This centralizes the unsafety of converting from UnsafeCell<T> to &mut T.

reem commented:
Ok, I think this should be ready to land now.

huonw commented on src/libstd/sync/mutex.rs:
````diff
@@ -369,32 +369,70 @@ impl StaticMutex {
+    /// ```
+    #[unstable(feature = "guard_map",
+               reason = "recently added, needs RFC for stabilization",
+               issue = "0")]
````
An issue should be opened and these numbers updated before this lands.

sfackler replied:
We're going to fold this into #27746 so that's the issue number to use here.

reem committed:
Add issue number to guard map methods.

reem commented:
Added the issue number and fixed the failing doctests, hopefully passes this time.

alexcrichton commented on src/libstd/sync/mutex.rs:
```diff
@@ -231,7 +231,7 @@ impl<T: ?Sized> Mutex<T> {
     pub fn try_lock(&self) -> TryLockResult<MutexGuard<T>> {
         if unsafe { self.inner.lock.try_lock() } {
-            Ok(try!(MutexGuard::new(&*self.inner, &self.data)))
+            Ok(try!(unsafe { MutexGuard::new(&*self.inner, &self.data) }))
```
With this new proliferation of `unsafe` blocks, can blocks like this just be merged with the one above? (e.g. have one surrounding block)

alexcrichton commented on src/libstd/sync/mutex.rs:
```diff
@@ -339,14 +339,14 @@ impl StaticMutex {
     pub fn lock(&'static self) -> LockResult<MutexGuard<()>> {
         unsafe { self.lock.lock() }
-        MutexGuard::new(self, &DUMMY.0)
+        unsafe { MutexGuard::new(self, &DUMMY.0) }
```
This is another case that would likely benefit from merging the two unsafe blocks

alexcrichton commented on src/libstd/sync/mutex.rs:
```diff
@@ -369,32 +369,70 @@ impl StaticMutex {
+               reason = "recently added, needs RFC for stabilization",
+               issue = "27746")]
+    pub fn map<U: ?Sized, F>(this: Self, cb: F) -> MutexGuard<'mutex, U>
+    where F: FnOnce(&'mutex mut T) -> &'mutex mut U {
```
ping on the style here of:

```rust
pub fn map<U: ?Sized, F>(this: Self, cb: F) -> MutexGuard<'mutex, U>
    where F: FnOnce(&'mutex mut T) -> &'mutex mut U 
{
    // ...
```

alexcrichton commented on src/libstd/sync/mutex.rs:
```diff
@@ -369,32 +369,70 @@ impl StaticMutex {
+        // We don't want to unlock the lock by running the destructor of the
+        // original lock, so just read the fields we need and forget it.
+        let poison = unsafe { ptr::read(&this.__poison) };
+        let lock = unsafe { ptr::read(&this.__lock) };
```
Can these unsafe blocks all be merged as well?

reem commented:
@alexcrichton updated for style changes both in mutex and rwlock

alexcrichton commented on src/libstd/sync/rwlock.rs:
```diff
@@ -178,7 +178,7 @@ impl<T: ?Sized> RwLock<T> {
     pub fn read(&self) -> LockResult<RwLockReadGuard<T>> {
         unsafe { self.inner.lock.read() }
-        RwLockReadGuard::new(&*self.inner, &self.data)
+        unsafe { RwLockReadGuard::new(&*self.inner, &self.data) }
```
Ah looks like the rwlock here needs similar treatment as a mutex

alexcrichton commented on src/libstd/sync/mutex.rs:
```diff
@@ -369,32 +377,72 @@ impl StaticMutex {
+    {
+        // Compute the new data while still owning the original lock
+        // in order to correctly poison if the callback panics.
+        let data = unsafe { ptr::read(&this.__data) };
```
This is an interesting aliasing situation where there are two valid `&mut` pointers to the same data (although one of them just isn't used). In the past we've received PRs to avoid this sort of situation, although I believe there is no technical reason currently why this would cause problems. I guess though that we can't reborrow for the outer lifetime, that's considered as moving out of the mutex guard?

This is when destructuring the mutex guard would be nice...

cc @Gankro, @bluss

Gankro replied:
Yeah I'm out of the aliasing model game. LLVM is so broken I don't even know what to believe anymore... and Rust doesn't even vaguely have something to point to.

reem replied:
I noticed this when refactoring to store an `&mut T` directly; you can't just reborrow since that marks the reference returned by the `cb` as borrowing the mutex, so you can't read out the lock/guard later.

Since the `&mut` in `this.__data` is not used while the other exists, I figured we would be in the clear, but it is a bit of a sketchy situation.

reem replied:
Note we actually don't really want to destructure the guard since we want panics in the `cb` to poison the mutex.

alexcrichton commented on src/libstd/sync/rwlock.rs:
```diff
@@ -412,48 +420,123 @@ impl StaticRwLock {
+               reason = "recently added, needs RFC for stabilization",
+               issue = "27746")]
+    pub fn map<U: ?Sized, F>(this: Self, cb: F) -> RwLockReadGuard<'rwlock, U>
+    where F: FnOnce(&'rwlock T) -> &'rwlock U {
```
looks like some realignment here as well

reem commented:
How would we feel about also adding `Ref::filter_map` equivalents?

alexcrichton commented:
In theory we're [slating `Ref::filter_map` for deprecation](https://github.com/rust-lang/rust/issues/27746) so it's fine to not add those as part of this PR.

@bors: r+ fc875b087cdd844fa61204db30c287f3c3179d92

Thanks!

bors commented:
:hourglass: Testing commit fc875b0 with merge 18c1781...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/7942), [auto-linux-32-opt](http://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/7963), [auto-linux-64-debug-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-debug-opt/builds/1209), [auto-linux-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/7903), [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/7913), [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/7902), [auto-linux-cross-opt](http://buildbot.rust-lang.org/builders/auto-linux-cross-opt/builds/1116), [auto-linux-musl-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-musl-64-opt/builds/3068), [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/7948), [auto-mac-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/7951), [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/7916), [auto-mac-ios-opt](http://buildbot.rust-lang.org/builders/auto-mac-ios-opt/builds/86), [auto-win-gnu-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-nopt-t/builds/2953), [auto-win-gnu-32-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/2969), [auto-win-gnu-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-nopt-t/builds/2962), [auto-win-gnu-64-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/2943), [auto-win-msvc-32-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/2176), [auto-win-msvc-64-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt/builds/2946)

reem commented:
@alexcrichton damn, I've discovered a safety hole in `MutexGuard::map` - it is not safe in combination with `Condvar`, since the "inner" borrow can be invalidated by another thread taking the lock during a `Condvar::wait` see: http://is.gd/RM038X.

reem commented:
This is not a problem with the `RwLock` methods, since the `RwLock` guards cannot be used with a `Condvar`.

reem commented:
It would be possible to provide a similar, but safe, API by defining a separate type for "mapped" mutex guards that can't be used with a `Condvar`.

alexcrichton commented:
Nice catch @reem! Certainly an interesting interaction between the two APIs...

pronebird commented:
Has it been nightly for like 10 years? 

https://doc.rust-lang.org/std/sync/struct.MutexGuard.html
